python_modules/dagster/dagster/_grpc/impl.py

```diff
@@ -283,24 +287,45 @@ def get_external_sensor_execution(
     definition = recon_repo.get_definition()
     sensor_def = definition.get_sensor_def(sensor_name)
 
-    with SensorEvaluationContext(
-        instance_ref,
-        last_completion_time=last_completion_timestamp,
-        last_run_key=last_run_key,
-        cursor=cursor,
-        repository_name=recon_repo.get_definition().name,
-    ) as sensor_context:
-        try:
-            with user_code_error_boundary(
-                SensorExecutionError,
-                lambda: "Error occurred during the execution of evaluation_fn for sensor "
-                "{sensor_name}".format(sensor_name=sensor_def.name),
-            ):
-                return sensor_def.evaluate_tick(sensor_context)
-        except Exception:
-            return ExternalSensorExecutionErrorData(
-                serializable_error_info_from_exc_info(sys.exc_info())
-            )
+    if isinstance(sensor_def, MultiAssetSensorDefinition):
+        with MultiAssetSensorEvaluationContext(
```

Comment:
nit: could get rid of the code dupe by using exitstack to handle the conditional contexts...

Reply:
could you explain a bit more? i'm not quite following

Reply:
You could do something like this:

```
with ExitStack() as stack:
    if isinstance(sensor_def, MultiAssetSensorDefinition):
        sensor_context = stack.enter_context(MultiAssetSensorEvaluationContext(...))
    else:
        sensor_context = stack.enter_context(SensorEvaluationContext(...))
    try:
        with user_code_error_boundary(...):
            return sensor_def.evaluate_tick(sensor_context)
    except Exception:
        return ExternalSensorExecutionErrorData(...)
```

Reply:
oh cool! i've never seen that before, that's super useful